### DISCUSSION TITLE
Sending only to designated connections

### DISCUSSION THREAD

**Discussion by whitwhoa** [uNetworking/uWebSockets#1463] [Q&A] [answered]
I'm almost embarrassed to ask this, as I know the answer is going to be obvious....but suppose I have 10 clients connected to the websocket server, but I only want to send a message to client 4 and 9. Where is the data stored that distinguishes between connections. I'm guessing there's a connection id somewhere? Or is there some built in component that I'm overlooking that handles this (like rooms)?

I see the following in one of the examples:
```
/* ws->getUserData returns one of these */
    struct PerSocketData {
        /* Fill with user data */
    };
...
.open = [](auto */*ws*/) {
            /* Open event here, you may access ws->getUserData() which points to a PerSocketData struct */

        },
```

Is PerSocketData struct prepopulated with this data, or are we to create this ourselves, and if so what values (and how do we obtain them) should we use to distinguish between users?

**Comment by ghost** [comment, marked as answer]
You are given a reference to a connection when said connection is established. How you mark or store this reference is up to you (do you expect some kind of JWT to log it in? Or do you expect some user name, etc?). You send using those references, as you wish.

**Reply by whitwhoa** [reply]
Is that reference shown in one of the examples? That's not `ws` is it? If so I'm looking at this the wrong way.

**Reply by ghost** [reply]
It's ws

**Reply by whitwhoa** [reply]
Got it. That makes sense now. For whatever reason I was thinking that `ws` was the server which held all connections and doing a `->send` was broadcasting to all connections. But, yeah looking at it all again I don't know how I could have drawn that conclusion, over(or under) thinking it I suppose. Thank you for your response! It is greatly appreciated :)

**Reply by ghost** [reply]
Ws.send is one-to-one
App.publish is one-to-many
Ws.publish is one-to-many, excluding itself
